test/int/featureBasedSuits/hitCountBreakpointTests.test.ts

```diff
@@ -0,0 +1,160 @@
+                hitCountCondition: '% 3'
+            });
+
+            await asyncRepeatSerially(2, () => incBtn.click());
+
+            await setStateBreakpoint.assertIsHitThenResumeWhen(() => incBtn.click(), `
+                increment [Counter.jsx] Line 17:12
+                onClick [Counter.jsx] Line 30:60
+                ${reactCounterAppBaseStack}`);
+
+            await incBtn.click();
+
+            await breakpoints.assertNotPaused();
+
+            await setStateBreakpoint.unset();
+        });
+
+        puppeteerTest("Hit count breakpoints = 3, = 4 and = 5 pause on the button's 3rd, 4th and 5th clicks", suiteContext, async (_context, page) => {
+            const incBtn = await page.waitForSelector('#incrementBtn');
+
+            const breakpoints = BreakpointsWizard.create(suiteContext.debugClient, reactTestSpecification);
+            const counterBreakpoints = breakpoints.at('Counter.jsx');
+
+            const setStateBreakpoint = await counterBreakpoints.hitCountBreakpoint({
+                lineText: 'this.setState({ count: newval });',
+                hitCountCondition: '= 3'
+            });
+
+            const setNewValBreakpoint = await counterBreakpoints.hitCountBreakpoint({
+                lineText: 'const newval = this.state.count + 1',
+                hitCountCondition: '= 5'
+            });
+
+            const stepInBreakpoint = await counterBreakpoints.hitCountBreakpoint({
+                lineText: 'this.stepIn();',
+                hitCountCondition: '= 4'
+            });
+
+            await asyncRepeatSerially(2, () => incBtn.click());
+
+            await setStateBreakpoint.assertIsHitThenResumeWhen(() => incBtn.click(), `
+                increment [Counter.jsx] Line 17:12
+                onClick [Counter.jsx] Line 30:60
+                ${reactCounterAppBaseStack}`);
+
+            await stepInBreakpoint.assertIsHitThenResumeWhen(() => incBtn.click(), `
+                increment [Counter.jsx] Line 18:12
+                onClick [Counter.jsx] Line 30:60
+                ${reactCounterAppBaseStack}`);
+
+            await setNewValBreakpoint.assertIsHitThenResumeWhen(() => incBtn.click(), `
+                increment [Counter.jsx] Line 16:27
+                onClick [Counter.jsx] Line 30:60
+                ${reactCounterAppBaseStack}`);
+
+            await incBtn.click();
+
+            await breakpoints.assertNotPaused();
+
+            await setStateBreakpoint.unset();
+            await setNewValBreakpoint.unset();
+            await stepInBreakpoint.unset();
+        });
+
+        puppeteerTest("Hit count breakpoints = 3, = 4 and = 5 set in batch pause on the button's 3rd, 4th and 5th clicks", suiteContext, async (_context, page) => {
+            const incBtn = await page.waitForSelector('#incrementBtn');
+
+            const breakpoints = BreakpointsWizard.create(suiteContext.debugClient, reactTestSpecification);
+            const counterBreakpoints = breakpoints.at('Counter.jsx');
```

Comment:
That is a good point.

I don't think we'll use the BreakpointsWizard in *every single test*. I do think we'll use BreakpointsWizard in most tests. I also think we might end up creating some other wizards or utilities for other things too.

What parameters would you pass to puppeteerSuite() so the "client" can initialize the BreakpointsWizard, or potentially another wizard/utility? Yet without forcing every single test to do it (Some loaded scripts tests won't require setting any breakpoints)?

Reply:
Hmm, for now I think we can hold off as we don't have a lot of different utilities, but in the future if we're going to be adding more stuff, it sounds to me like a good place for a builder pattern, where in creation of the suite, we could add modules that would be able to add functionality to the tests within that suite.

Something like:


    suiteBuilder()
        .add(new PuppeteerThing())
        .add(new BreakpointWizard())
        .build( (context) => {
             //test code here, and utils accessible from the context?
        });

Might have to think about how to make that type-safe, but it would make things much more composeable than the current `puppeteerSuite`, and for each suite we could add only the functionality required for those tests.

Reply:
That sounds like a great idea! Let's keep an eye of how the tests evolve, and look into this after we know if we need more utilities/wizards or not...